kratos/sources/registry.cpp

```diff
@@ -80,6 +81,16 @@ namespace
         }
     }
 
+    void Registry::SetCurrentSource(std::string const & rCurrentSource)
+    {
+        mCurrentSource = rCurrentSource;
+    }
+
+    std::string Registry::GetCurrentSource()
+    {
+        return mCurrentSource;
+    }
```

Comment:
Should be fine, registering an application should not be done in parallel and the registry itself relies on being static.

In this regard we may want to implement Registry as a singleton and have a thread lock in the setters/getters. Ping @pooyan-dadvand.

Reply:
Thinking more about it, I think it would be better to first register the application in the registry and keep the entry as an static member of the application itself:

``` c++
static RegistryItem& mRegisteredApplication = Registry::AddItem<RegistryItem>("FluidDynamics");
```

Then we can use this name everywhere using the `Application` base class member function `GetApplicationName` (you may find better names for them to be coherent also in the core)

In the core we can use `KratosCore` or `Core` as the name.

Then we can add all items not only to their type but also as a reference under this branch (optional) and have a `deregistry_list` with all added items to be removed:
``` text
application_name -- elements [ ] // optional but can be useful to find the source of certain component
                 |- conditions 
                 |- ...
                 |- deregsitery_list -- elements // only the ones that we added
                                     |- conditions   // only what we added
                                     |- geometries  // ....
                                     |- .....
``` 

With all of this in place, the deregistry function can be implemented in a generic way in the base class or as a utility

Reply:
I think I get the idea but I have a comment:

> Then we can use this name everywhere using the Application base class member function `GetApplicationName` (you may find better names for them to be coherent also in the core)

Essentially, this information is already available in https://github.com/KratosMultiphysics/Kratos/blob/master/kratos/includes/kratos_application.h#L155. The motivation for adding a `mCurrentSource` is to get access to this information outside the application context (for example, for registering preconditioners https://github.com/KratosMultiphysics/Kratos/blob/master/kratos/factories/standard_preconditioner_factory.cpp#L47-L50 )

With that being said, what we can do is to store this "current application" inside the registry itself instead of being a static member. That will also solve the thread-safety concerns. Modifying a bit your proposal would be like:

```text
application_context --> @app_name
application_name -- elements [ ] // optional but can be useful to find the source of certain component
                 |- conditions 
                 |- ...
                 |- deregsitery_list -- elements // only the ones that we added
                                     |- conditions   // only what we added
                                     |- geometries  // ....
                                     |- .....
``` 

And then the registering process would be like:

```C++
auto & app_context = Registry::GetItem<RegistryItem>("ApplicationContext");
if (Registry::HasItem("elements."+app_context+"."+name)) ) // or whatever access.
```

Reply:
Btw I remember that we agreed that registry keys should start with the component name and then the application name (`process.all.xxxx`,  `process.FluidDynamics.yyyyy`), so it should be:

```text
application_context --> @app_name
elements 
            |- application_name
                |- deregsitery_list
conditions 
            |- application_name
                |- deregsitery_list
```